### PR TITLE
`sparsezoo.analyze`

### PR DESCRIPTION
**Pull request by rahul-tuli**
# `sparsezoo.analyze`

Current PR represents the feature branch for `sparsezoo.analyze` utility

This utility can be used to run `ModelAnalysis` on ONNX models, and prints
relevant summary to screen, if a `--save` option is given then the 
complete analysis is written to the specified file

## Usage

```bash
Usage: sparsezoo.analyze [OPTIONS] MODEL_PATH

  Model analysis for ONNX models.

  MODEL_PATH: can be a SparseZoo stub, or local path to a deployment-directory
  or ONNX model

  Examples:

  - Run model analysis on resnet

      sparsezoo.analyze ~/models/resnet50.onnx

Options:
  --save FILE  Path to a yaml file to write results to, note: file will be
               overwritten if exists
  --help       Show this message and exit.
```

Examples:
[UPDATED OUTPUT]

**COMMAND:**
```bash
sparsezoo3.9) 🥃 sparsezoo sparsezoo.analyze "zoo:cv/classification/resnet_v1-50/pytorch/sparseml/imagenet/pruned85_quant-none-vnni"
```

**OUTPUT:**
```bash
INFO:root:Downloading files from SparseZoo: 'zoo:cv/classification/resnet_v1-50/pytorch/sparseml/imagenet/pruned85_quant-none-vnni'
INFO:root:Starting Analysis ...
INFO:root:Analysis complete, collating results...
MODEL: zoo:cv/classification/resnet_v1-50/pytorch/sparseml/imagenet/pruned85_quant-none-vnni

PARAMETERS:
             Total  Percent Total %  Sparsity %  FP32 Precision %  INT8 Precision %
Weight  25502912.0            99.89       79.22              8.03             91.97
Bias       27560.0             0.11        0.00              3.63             96.37
Total   25530472.0           100.00       79.13              8.03             91.97

PARAMETERIZED OPS:
                  Total  Percent Total %  Sparsity %  FP32 Precision %  INT8 Precision %
QLinearConv  23481472.0            91.97       86.04              0.00            100.00
Gemm          2049000.0             8.03        0.00            100.00              0.00
Total        25530472.0           100.00       79.13              8.03             91.97

NON PARAMETERIZED OPS:
                   Total  Percent Total %  Sparsity %  FP32 Precision %  INT8 Precision %
Add                 16.0            18.39         0.0             100.0               0.0
Reshape              1.0             1.15         0.0             100.0               0.0
Unsqueeze            1.0             1.15         0.0             100.0               0.0
QuantizeLinear      17.0            19.54         0.0             100.0               0.0
Concat               1.0             1.15         0.0             100.0               0.0
Relu                13.0            14.94         0.0             100.0               0.0
GlobalAveragePool    1.0             1.15         0.0             100.0               0.0
Shape                1.0             1.15         0.0             100.0               0.0
DequantizeLinear    33.0            37.93         0.0             100.0               0.0
MaxPool              1.0             1.15         0.0             100.0               0.0
Softmax              1.0             1.15         0.0             100.0               0.0
Gather               1.0             1.15         0.0             100.0               0.0
Total               87.0           100.00         0.0             100.0               0.0

SUMMARY:
Number of Parameters: 25530472
Number of Operations: 25530559
Weight Sparsity %: 79.22
Quantized Parameterized Ops %: 91.97
Quantized Non-Parameterized Ops %: 0.0
```

- [X] Add Tests #264 
- [X] Update Summary #269 
- [X] Update Summary to be more Human Readable #273 
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203828991992718

### CODE REVIEW

**Review by KSGulin — DISMISSED** [review]
Super clean. Excited to have this tool

**Review by KSGulin — APPROVED** [review]

File: src/sparsezoo/analysis_cli.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    summary = _get_analysis_summary(analysis=analysis)
+    summary["model_path"] = model_path
+    pp.pprint(summary)
+
+    if save:
+        LOGGER.info(f"Writing results to {save}")
+        analysis.yaml(file_path=save)
+
+
+def _get_analysis_summary(analysis):
```
Comment (bfineran) [inline src/sparsezoo/analysis_cli.py:114]: let's move this to a `.summary()`  helper function of `ModelAnalysis`

**Review by bfineran — COMMENTED** [review]
LGTM pending coment

File: src/sparsezoo/analysis_cli.py (hunk 1 above)
Comment (rahul-tuli) [reply src/sparsezoo/analysis_cli.py:114]: Oooooooh nice, yes!

**Review by bfineran — APPROVED** [review]

**Review by markurtz — CHANGES_REQUESTED** [review]
@rahul-tuli I've added more details into the PRD to prescribe the UX for presenting the results. The JSON works, but we'd like to format it into a more readable format. Attaching a screenshot from that PRD

**Comment by markurtz** [comment]
Screen Shot 2023-02-15 at 8 19 20 AM

**Comment by rahul-tuli** [comment]
> Screen Shot 2023-02-15 at 8 19 20 AM

@markurtz Updated!

**Review by KSGulin — APPROVED** [review]
LGTM! Looks like there's a merge conflict to resolve

**Review by KSGulin — APPROVED** [review]

**Review by markurtz — CHANGES_REQUESTED** [review]
@rahul-tuli this is still printing out as a dict rather than the easier to read and more compact table format

**Review by markurtz — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**